src/Build/Evaluation/IEvaluatorData.cs

```diff
@@ -263,7 +263,7 @@ List<ProjectItemElement> EvaluatedItemElements
         /// <summary>
         /// Sets a property which does not come from the Xml.
         /// </summary>
-        P SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved);
+        P SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved, bool isEnvVar = false);
```

Comment:
My suggestion is to add environment variable reads and uninitialized property reads to Project as well, not just ProjectInstance.

In certain cases, ProjectInstances are created from Project instances via [https://github.com/microsoft/msbuild/blob/master/src/Build/Definition/Project.cs#L2696](https://github.com/microsoft/msbuild/blob/master/src/Build/Definition/Project.cs#L2696)

Quickbuild and VS do this. @dfederm, would qb use this information ?

Reply:
QuickBuild mostly just uses `ProjectInstance`. We use a few small pieces of `Project` when we need unevaluated values, but only when necessary. We prefer using `ProjectInstance` when possible.

Reply:
Do you initialize the ProjectInstance from a Project?